awx/main/tasks/facts.py

```diff
@@ -34,30 +42,32 @@ def start_fact_cache(hosts, destination, log_data, timeout=None, inventory_id=No
     if timeout is None:
         timeout = settings.ANSIBLE_FACT_CACHE_TIMEOUT
 
-    if isinstance(hosts, QuerySet):
-        hosts = hosts.iterator()
-
     last_filepath_written = None
+
     for host in hosts:
-        if (not host.ansible_facts_modified) or (timeout and host.ansible_facts_modified < now() - datetime.timedelta(seconds=timeout)):
+        if not host.ansible_facts_modified or (timeout and host.ansible_facts_modified < now() - datetime.timedelta(seconds=timeout)):
             continue  # facts are expired - do not write them
+
         filepath = os.sep.join(map(str, [destination, host.name]))
         if not os.path.realpath(filepath).startswith(destination):
             system_tracking_logger.error('facts for host {} could not be cached'.format(smart_str(host.name)))
             continue
+
         try:
             with codecs.open(filepath, 'w', encoding='utf-8') as f:
                 os.chmod(f.name, 0o600)
                 json.dump(host.ansible_facts, f)
                 log_data['written_ct'] += 1
                 last_filepath_written = filepath
+            hosts_cached.append(CachedHost(name=host.name, inventory=host.inventory, ansible_facts=host.ansible_facts))
```

Comment:
I still expect that this won't work. Again, when going through the `for host in hosts:` loop, many hosts will hit the `continue` statement before here, meaning they won't get into the list. Really, if you run this for the first time, **no** hosts will be in the list, and I legitimately think it will be `[]`. This is not what you want. For this to work as the variable of iteration in `finish_fact_cache`, it needs to basically contain all hosts in the inventory (of the section of inventory for sliced jobs).

And this is a very creative approach to create a new class, but keeping all ansible_facts in memory is too much. You also shouldn't need them, they shouldn't be used in `finish_fact_cache`.